semanticdb/semanticdb3/javacp.md

```diff
@@ -0,0 +1,480 @@
+Notes:
+* Class members must appear in the following order:
+  * non-static members first, following the same order as they appear in the original source
+  * static members secondly, following the same order as they appear in the original source
+* A Java class maps to a single symbol with type `ClassInfoType` including all static and non-static members.
+  This departs from the Scala compiler internal representation of Java classes where non-static members
+  are grouped under a `CLASS` symbol and static members are grouped under an `OBJECT` symbol.
+* Supported properties for `CLASS` symbols are
+  * `FINAL` set for all final classes
+  * `ABSTRACT` set for all final classes
+  * `STATIC` set for static classes
+  * `ENUM` set for enum declarations
+
+**Enum types** [\[84\]][84] are represented by a single symbol with the `CLASS` kind.
+
+```java
+package a;
+
+public enum Coin {
+  PENNY, NICKEL
+}
+```
+
+<table>
+  <tr>
+    <td><b>Definition</b></td>
+    <td width="275px"><b>Symbol</b></td>
+    <td><b>Kind</b></td>
+    <td><b>Signature</b></td>
+  </tr>
+  <tr>
+    <td><code>Coin</code></td>
+    <td><code>a.Coin#</code></td>
+    <td><code>CLASS</code></td>
+    <td><code>ClassInfoType(List(), List(&lt;Enum&lt;Coin&gt;&gt;), List(a.Coin#PENNY., a.Coin#NICKEL., ))</code></td>
+  </tr>
+  <tr>
+    <td><code>Coin</code></td>
+    <td><code>a.Coin#PENNY.</code></td>
+    <td><code>FIELD</code></td>
+    <td><code>TypeRef(None, &lt;a.Coin#&gt;), List())</code></td>
+  </tr>
+  <tr>
+    <td><code>Coin</code></td>
+    <td><code>a.Coin#NICKEL.</code></td>
+    <td><code>FIELD</code></td>
+    <td><code>TypeRef(None, &lt;a.Coin#&gt;), List())</code></td>
+  </tr>
+</table>
+
+Notes:
+* Enum types follow the same rules as class declarations.
+* Enum type classes and enum fields must have the property `ENUM` set.
+* Enum fields have the type of their enclosing class.
+
+**Interface declarations** [\[77\]][77] are represented by a single symbol like classes but with the `INTERFACE` kind.
+Concretely, the differences between interface symbols and class symbols are:
+* Interface symbols don't any properties.
+* Traits don't have symbols.
+
+**Method declarations** [\[82\]][82] are represented by a single symbol with the `METHOD` kind and one symbol for each method parameter with kind `PARAMETER`.
+Notes:
+* When compiled with `-parameters` , the name of method parameters matches their name written in source. Otherwise, parameters have the name `paramN` where `N` is the index of that given parameter starting at index 0.
+
+<a name="java-accessibility"></a>
+#### Accessibility
+
+```protobuf
+message Accessibility {
+  enum Tag {
+    UNKNOWN_ACCESSIBILITY = 0;
+    PRIVATE = 1;
+    PRIVATE_THIS = 2;
+    PRIVATE_WITHIN = 3;
+    PROTECTED = 4;
+    PROTECTED_THIS = 5;
+    PROTECTED_WITHIN = 6;
+    PUBLIC = 7;
+  }
+  Tag tag = 1;
+  string symbol = 2;
+}
+```
+
+In Java, [Accessibility](#accessibility) represents accessibility of definitions.
```

Comment:
Are there parts of the spec that describe this?

Reply:
Yes! Updated to link the section on "Access control" https://docs.oracle.com/javase/specs/jls/se8/html/jls-6.html#jls-6.6